File: MARATONA DE PROGRAMAÇÃO SBC 2017/Escala Musical.py
tabela = [2,2,1,2,2,2,1]
mapa=dict()
notas= ["do","do#","re","re#","mi","fa","fa#","sol","sol#","la","la#","si"]
escala = [[0 for x in range(12)] for y in range(12)] 
tam=12
for i in range(0,12):
    k=0
    j=i
    proxNota=i
    flag=True
    while(True):
        if j==i and flag:
            flag=False
        elif j==i and not(flag):
            break;
        if j==proxNota:
            escala[i][j]=1
            proxNota+=tabela[k]
            k+=1
        j+=1
        if j==12:
            j=0
            proxNota-=12

n = int(input())
for i in range(0,n):
    aux = int(input())
    aux-=1
    if aux>=12:
        aux=aux%12
    for j in range(0,12):
        if escala[j][aux]==1:
            mapa[j] = mapa.get(j, 0) + 1

flag = False
for key, value in mapa.items():
    if value==n:
        print(notas[key])
        flag=True
        break
if not(flag):
    print("desafinado")

File: MARATONA DE PROGRAMAÇÃO SBC 2017/Escala Musical.cpp
#include <bits/stdc++.h>

using namespace std;

int tabela [] = {2,2,1,2,2,2,1};
map<int,int> mapa;
string notas[]= {"do","do#","re","re#","mi","fa","fa#","sol","sol#","la","la#","si"};
int escala [12][12];
int main()
{
    memset(escala,0,sizeof(escala));
    for(int i=0; i<12; i++)
    {
        int k=0;
        int j=i;
        int proxNota=i;
        bool flag=true;
        while(true)
        {

            if(j==i && flag)
                flag=false;
            else if(j==i && !flag)
                break;
            if(j==proxNota)
            {
                escala[i][j]=1;
                proxNota+=tabela[k++];
            }
            j++;
            if(j==12)
            {
                j=0;
                proxNota-=12;
            }
        }
    }

    int n;
    cin>>n;
    for(int i=0; i<n; i++)
    {
        int aux;
        cin>>aux;
        aux-=1;
        if(aux>12)
            aux=aux%12;
        for(int j=0; j<12; j++)
        {
            if(escala[j][aux]==1)
            {
                if(mapa.find(j)==mapa.end())
                    mapa[j]=0;
                mapa[j]=mapa[j]+1;
            }
        }
    }
    bool flag = false;
    for(map<int,int>::iterator it=mapa.begin(); it!=mapa.end(); it++)
    {
        if(it->second==n)
        {
            cout<<notas[it->first]<<endl;
            flag=true;
            break;
        }
    }
    if(!flag)
        cout<<"desafinado"<<endl;
    return 0;
}
